compliance/virtualmachine/relay.go

```diff
@@ -0,0 +1,236 @@
+	// replaced with a sanity check.
+	indexReport.VsockCid = strconv.Itoa(int(vsockCID))
+
+	if err = sendReportToSensor(ctx, indexReport, sensorClient); err != nil {
+		return errors.Wrapf(err, "sending report to sensor (vsock CID: %d)", vsockCID)
+	}
+
+	log.Debugf("Finished handling vsock connection from %s", conn.RemoteAddr())
+
+	return nil
+}
+
+func isRetryableGRPCError(err error) bool {
+	grpcErr, ok := status.FromError(err)
+	if !ok {
+		return false
+	}
+	code := grpcErr.Code()
+	switch code {
+	case codes.DeadlineExceeded:
+		return !errors.Is(err, context.Canceled)
+	case codes.Unavailable, codes.ResourceExhausted, codes.Internal:
+		return true
+	default:
+		return false
+	}
+}
+
+func parseIndexReport(data []byte) (*v1.IndexReport, error) {
+	report := &v1.IndexReport{}
+
+	if err := proto.Unmarshal(data, report); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling data")
+	}
+	return report, nil
+}
+
+func readFromConn(conn net.Conn, maxSize int, timeout time.Duration) ([]byte, error) {
+	if err := conn.SetReadDeadline(time.Now().Add(timeout)); err != nil {
+		return nil, errors.Wrap(err, "setting read deadline on connection")
+	}
+
+	// Add 1 to the limit so we can detect oversized data. If we used exactly maxSize, we couldn't tell the difference
+	// between a valid message of exactly maxSize bytes and an invalid message that's larger than maxSize (both would
+	// read maxSize bytes). With maxSize+1, reading more than maxSize bytes means the original data was too large.
+	limitedReader := io.LimitReader(conn, int64(maxSize+1))
+	data, err := io.ReadAll(limitedReader)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading data from vsock connection")
+	}
+
+	if len(data) > maxSize {
+		return nil, errors.Errorf("data size exceeds the limit (%d bytes)", maxSize)
+	}
+
+	return data, nil
+}
+
+func sendReportToSensor(ctx context.Context, report *v1.IndexReport, sensorClient sensor.VirtualMachineIndexReportServiceClient) error {
+	log.Debugf("Sending index report to sensor (vsockCID: %s)", report.VsockCid)
+
+	req := &sensor.UpsertVirtualMachineIndexReportRequest{
+		IndexReport: report,
+	}
+
+	// Considering a timeout of 5 seconds and 10 tries with exponential backoff, the maximum time spent in this function
+	// is around 1 min 40 s. Given that each virtual machine sends an index report every 4 hours, these retries seem
+	// reasonable and are unlikely to cause issues.
+	err := retry.WithRetry(func() error {
+		sendToSensorCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		defer cancel()
+		resp, err := sensorClient.UpsertVirtualMachineIndexReport(sendToSensorCtx, req)
+
+		if resp != nil && !resp.Success {
+			// This can't happen as of this writing (Success is only false when an error is returned) but is
+			// theoretically possible, let's add retries too.
+			if err == nil {
+				err = retry.MakeRetryable(errors.New("Sensor failed to handle virtual machine index report"))
```

Comment:
I think, it maybe useful to print the resp.  Landing in this execution branch may be rare and it would be good to have all troubleshooting info at hand.

Reply:
Good idea, thanks. Done: https://github.com/stackrox/stackrox/pull/16929/commits/8053433cd6ecb21bdf120c0202f6c04598c27145